shale/src/compact.rs

```diff
@@ -654,14 +655,16 @@ impl<T: Storable + Send + Sync + Debug + 'static, M: CachedStore + Send + Sync>
         let payload_size = inner
             .get_header(ptr - CompactHeader::MSIZE as usize)?
             .payload_size;
+        let obj = self.obj_cache.put(inner.get_data_ref(ptr, payload_size)?);
+        let cache = &self.obj_cache;
 
-        Ok(inner.obj_cache.put(inner.get_data_ref(ptr, payload_size)?))
+        Ok(ObjRef::new(Some(obj), cache))
     }
 
     fn flush_dirty(&self) -> Option<()> {
         let mut inner = self.inner.write().unwrap();
         inner.header.flush_dirty();
-        inner.obj_cache.flush_dirty()
+        self.obj_cache.flush_dirty()
```

Comment:
The guard now lives too long. Should be:
```
self.inner.write().unwrap().header.flush_dirty()
self.obj_cache.flush_dirty()
```

Reply:
@rkuris, I added a comment here to keep these actions in-sync. Let me know if you still think I should change this